exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/grpc/GrpcExporterBuilder.java

```diff
@@ -5,39 +5,211 @@
+  public GrpcExporterBuilder<T> setCompression(String compressionMethod) {
+    this.compressionEnabled = true;
+    return this;
+  }
 
-  GrpcExporter<T> build();
+  public GrpcExporterBuilder<T> setTrustedCertificates(byte[] trustedCertificatesPem) {
+    this.trustedCertificatesPem = trustedCertificatesPem;
+    return this;
+  }
+
+  public GrpcExporterBuilder<T> setClientTls(byte[] privateKeyPem, byte[] certificatePem) {
+    this.privateKeyPem = privateKeyPem;
+    this.certificatePem = certificatePem;
+    return this;
+  }
+
+  public GrpcExporterBuilder<T> addHeader(String key, String value) {
+    headers.put(key, value);
+    return this;
+  }
+
+  public GrpcExporterBuilder<T> setRetryPolicy(RetryPolicy retryPolicy) {
+    this.retryPolicy = retryPolicy;
+    return this;
+  }
+
+  public GrpcExporterBuilder<T> setMeterProvider(MeterProvider meterProvider) {
+    this.meterProvider = meterProvider;
+    return this;
+  }
+
+  public GrpcExporter<T> build() {
+    if (grpcChannel != null) {
+      return new UpstreamGrpcExporterFactory().buildWithChannel((Channel) grpcChannel);
+    }
+
+    OkHttpClient.Builder clientBuilder =
+        new OkHttpClient.Builder().dispatcher(OkHttpUtil.newDispatcher());
+
+    clientBuilder.callTimeout(Duration.ofNanos(timeoutNanos));
+
+    if (trustedCertificatesPem != null) {
+      try {
+        X509TrustManager trustManager = TlsUtil.trustManager(trustedCertificatesPem);
+        X509KeyManager keyManager = null;
+        if (privateKeyPem != null && certificatePem != null) {
+          keyManager = TlsUtil.keyManager(privateKeyPem, certificatePem);
+        }
+        clientBuilder.sslSocketFactory(
+            TlsUtil.sslSocketFactory(keyManager, trustManager), trustManager);
+      } catch (SSLException e) {
+        throw new IllegalStateException(
+            "Could not set trusted certificates, are they valid X.509 in PEM format?", e);
+      }
+    }
+
+    String endpoint = this.endpoint.resolve(grpcEndpointPath).toString();
+    if (endpoint.startsWith("http://")) {
+      clientBuilder.protocols(Collections.singletonList(Protocol.H2_PRIOR_KNOWLEDGE));
+    } else {
+      clientBuilder.protocols(Arrays.asList(Protocol.HTTP_2, Protocol.HTTP_1_1));
+    }
+
+    Headers.Builder headers = new Headers.Builder();
+    this.headers.forEach(headers::add);
+
+    headers.add("te", "trailers");
+    if (compressionEnabled) {
+      headers.add("grpc-encoding", "gzip");
+    }
+
+    if (retryPolicy != null) {
+      clientBuilder.addInterceptor(
+          new RetryInterceptor(retryPolicy, OkHttpGrpcExporter::isRetryable));
+    }
+
+    return new OkHttpGrpcExporter<>(
+        exporterName,
+        type,
+        clientBuilder.build(),
+        meterProvider,
+        endpoint,
+        headers.build(),
+        compressionEnabled);
+  }
+
+  // Use an inner class to ensure GrpcExporterBuilder does not have classloading dependencies on
+  // upstream gRPC.
```

Comment:
@anuraaga I think there is a bug here.  At the time UpstreamGrpcExporterFactory inner class is called on line 133, the current ClassLoader is the `AgentClassLoader` (io.opentelemetry.javaagent.bootstrap.AgentClassLoader), which does not have access to `io.grpc.Codec`.  `io.grpc.Codec` may be available in an ExtensionClassloader, but not here when line 133 is called.  Thus, everything blows up with a `NoClassDefFoundError`.

Reply:
I'm speaking in the context of the [java instrumentation agent](https://github.com/open-telemetry/opentelemetry-java-instrumentation) and an extension that tries to provide a `Channel`.

Reply:
hi @JesseEstum! can you open an issue for this? otherwise likely to get lost on this old PR

Reply:
Good point, @trask .  After reflecting over the weekend, I realized this would be a non-issue if there was a way to control the class path of the `AgentClassLoader` and perhaps I just "wasn't doing it right".

As such, I opened https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/9029 as an alternative option.  I'll let that one play out first then if necessary open an issue against this repo.